### ISSUE TITLE
C++ runtime is not bundled

### ISSUE THREAD

**Issue by stschake** [probonopd/linuxdeployqt#155]
Running my linuxdeployqt created AppImage on an older Ubuntu 14.04 gives the following runtime error:

    relocation error: ./My.AppImage: symbol _ZNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEE9_M_createERmm, version GLIBCXX_3.4.21 not defined in file libstdc++.so.6 with link time reference

14.04 seems to ship with a C++ runtime < version GLIBCXX_3.4.21, causing the error. However, I didn't expect it to occur in the first place because I figured linuxdeployqt would bundle all the system libs that my executable depends on, including the C++ runtime. However, looking at the extracted AppImage, it's not present while other system libraries are.

**Comment by TheAssassin** [comment]
Basic libraries such as lib[std]c++ are not supposed to be bundled, they are commonly available. This is the reason why the library is not in the AppImage.

As https://github.com/AppImage/AppImages/blob/master/excludelist#L31 shows, actually, excluding the library is supposed to solve the exact same problem you have there. That's a really interesting fact. We'll have to investigate that.

**Comment by probonopd** [comment]
Your application, compiled on that system, will need at least libstdc++ 3.4.21.

**Comment by stschake** [comment]
Right, I can understand setting a reasonable baseline, mandating something like glibc - where bundling it isn't feasible anyway due to the demands on the underlying operating system.

Is there a reason that precludes bundling libstdc++? I don't understand the rationale given in the exclude list. With >= GCC 5.1 (which introduced 3.4.21) now 2 years old and shipped by default in Ubuntu, Debian and other major distributions this is going to be a rather common issue for Qt applications. Having it only blow up at runtime is rather unfortunate.

**Comment by probonopd** [comment]
We consider libstdc++ to be present on all target systems. Hence there is no need to bundle it (and bloat the AppImage) if we can compile the software against an old enough version of libstdc++. If that is not feasible for some reason, then I don't see a strong reason why you couldn't bundle your own private copy. Just manually copy it in between 2 runs of `linuxdeployqt`.

**Comment by YiGene** [comment]
i got similar issue. i was hoping linuxdeployqt tool can help to resolve it but seems it's not..
/Application-x86_64.AppImage
./Application-x86_64.AppImage: relocation error: /tmp/.mount_ApplicLXADsD/lib/libc.so.6: symbol _dl_find_dso_for_object, version GLIBC_PRIVATE not defined in file ld-linux-x86-64.so.2 with link time reference

so i have to use the normal way to deploy the files and run with this way.
./ld-linux-x86-64.so.2 --library-path /libs /yourapp

**Comment by cvuchener** [comment]
Compiling C++14 on Ubuntu 14.04 requires gcc-5 (I am using the package from https://launchpad.net/~ubuntu-toolchain-r/+archive/ubuntu/test), but this updates libstdc++.so. I don't know a way to build and distribute a C++14 program without either bundling libstdc++.so or linking it statically.

**Comment by TheAssassin** [comment]
You will have to build your software on a later system (16.04) then, or try to use C++11 only.

**Comment by cvuchener** [comment]
The issue is not building, the ppa makes that easy. The issue is running the resulting appimage on other Ubuntu 14.04 without the ppa or Debian Jessie.

**Comment by TheAssassin** [comment]
Well, that won't be possible without some dirty hacks.

**Comment by TheAssassin** [comment]
@cvuchener actually, @probonopd recently posted some link to a piece of software making what you need possible. I didn't test it, and I don't remember the name... but it might help you.

**Comment by probonopd** [comment]
Please see https://github.com/AppImage/AppImageKit/wiki/Creating-AppImages#libstdcso6, especially the linked https://github.com/darealshinji/AppImageKit-checkrt/.

**Comment by cvuchener** [comment]
Thanks.

How should I use that with linuxdeployqt? If I copy the patched AppRun before running linuxdeployqt, it is replaced with a symlink to my executable.

**Comment by TheAssassin** [comment]
@cvuchener please ask the author of the software for more information.

In general, if you need to use a custom AppRun with linuxdeployqt, you need to run linuxdeployqt first, delete the AppRun symlink it creates, and put in your file.

**Comment by probonopd** [comment]
Here is an example where I am using it:

https://github.com/probonopd/audacity/blob/AppImage/.travis.yml

The key lines are:

```
  - # Workaround to increase compatibility with older systems; see https://github.com/darealshinji/AppImageKit-checkrt for details
  - mkdir -p appdir/usr/optional/ ; wget -c https://github.com/darealshinji/AppImageKit-checkrt/releases/download/continuous/exec-x86_64.so -O ./appdir/usr/optional/exec.so
  - mkdir -p appdir/usr/optional/libstdc++/ ; cp /usr/lib/x86_64-linux-gnu/libstdc++.so.6 ./appdir/usr/optional/libstdc++/
```
